blog/content/post/2021-12-14-secure-jtag-on-imx-soc.md

```diff
@@ -0,0 +1,280 @@
+from Olimex.
+
+- **System image**: Any Linux image for your i.MX6 platform with kernel module
+`nvmem-imx-ocotp` for modify OTP memory.
+
+## How it works
+
+SoC's from i.MX family offers JTAG in three modes
+
+- **No debug** - security on the highest level, but after setting that
+development and debugging via JTAG port will be unavailable permanently on this
+device
+
+- **Secure JTAG** - mode described in this article, the best compromise between
+security and development possibilities - only person with access key can use
+JTAG
+
+- **Enabled JTAG** - default mode, enabled for anyone who has physical access to
+JTAG pins or even only device (is a chance to connect to the JTAG pins if the
+connector isn't soldered)
+
+Secure JTAG mode is based on challenge/response mechanism. SoC needs to save a
+unique challenge key during manufacturing. User can generate their response
+56-bit key and burn it in JTAG fuse named `SJC_RESP`. While trying to access the
+JTAG port, SJC gives a challenge key. We can simplify this to the JTAG
+controller ID. Now user should pass the response key. It is compared with the
+response stored in SoC fuse bits. If keys are the same, JTAG is enabled. Below,
+you can see diagram described this mechanism.
+
+![Secure JTAG - how it works](/img/secure_jtag.png)
+
+JTAG mode can be set in `JTAG_SMODE` fuse by values
+
+| Mode        | JTAG_SMODE |
+|-------------|------------|
+| JTAG enable | 0x0        |
+| JTAG secure | 0x1        |
+| No debug    | 0x2        |
+
+These fuses still can be overwritten until we do not block this possibility -
+the response key can be locked by writing `0x1` to `SJC_RESP_LOCK`, and JTAG
+mode with all other fuses from `BOOT` group by writing `0x3` to `BOOT_CFG_LOCK`.
+Now they should be protected from writing and overriding. There is an option to
+set this to locking only override - look at the reference manual or [application
+note](https://usermanual.wiki/m/bb676916d740bdd5d4e8ba43c1ba41673c242f2cefc59f03e014ea7314314f62.pdf)
+for more information about this feature.
+
+## Setting up
+
+### Using cruible
+
+Get the latest prebuild release from
+[official project page](https://github.com/f-secure-foundry/crucible/releases/download/v2021.05.03/crucible)
+(on the day of writing this article it is `v2021.05.03`). You can also download
+source code and build it by yourself - this process is described in
+[readme](https://github.com/f-secure-foundry/crucible#installing). This program
+doesn't require any specific installation procedure - just remember to give
+executable access to this binary file.
+
+Run `cruible` on your target and check that you have access to fuse bits from
+userspace
+```
+# ./crucible -m IMX6UL -r 1 -b 16 read JTAG_SMODE
+soc:IMX6UL ref:1 otp:JTAG_SMODE op:read addr:0x18 off:22 len:2 val:0x0
+```
+
+> Note: As mentioned at the start, a JTAG on any i.MX SoC can be fused but if
+you try to burn fuses on i.MX8MM there can be problems with availability of
```

Comment:
It is a part of our research about the possibilities of JTAG fusing on i.MX8M-Mini - the article is about fusing i.mx6 but we mentioned that JTAG mode can be set in a similar way on different imx platforms - so it can be useful tip for someone who want to secure JTAG on different SoC than imx6

Reply:
I googled i.MX8MM and could not find what does it mean, we shoudl use .MX8M-Mini